projects/lesson-5/runtime/src/kitties.rs

```diff
@@ -0,0 +1,442 @@
+			
+
+			Self::do_transfer(&sender, to, kitty_id)
+		}
+		// 作业：实现 transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex)
+		// 使用 ensure! 来保证只有主人才有权限调用 transfer
+		// 使用 OwnedKitties::append 和 OwnedKitties::remove 来修改小猫的主人
+
+		pub fn buy_kitty(origin,  kitty_id: T::KittyIndex, max_price: BalanceOf<T>) -> Result{
+			let sender = ensure_signed(origin)?;
+			Self::do_buy_kitty(&sender, kitty_id, max_price)
+		}
+		pub fn set_price(origin,  kitty_id: T::KittyIndex, price : BalanceOf<T>) -> Result{
+			let sender = ensure_signed(origin)?;
+			Self::do_set_price(&sender, kitty_id, price)
+
+		}
+	}
+}
+
+impl<T: Trait> OwnedKitties<T> {
+	
+	fn read_head(account: &T::AccountId) -> KittyLinkedItem<T> {
+ 		Self::read(account, None)
+ 	}
+
+	fn write_head(account: &T::AccountId, item: KittyLinkedItem<T>) {
+ 		Self::write(account, None, item);
+ 	}
+
+	fn read(account: &T::AccountId, key: Option<T::KittyIndex>) -> KittyLinkedItem<T> {
+ 		<OwnedKitties<T>>::get(&(account.clone(), key)).unwrap_or_else(|| KittyLinkedItem {
+ 			prev: None,
+ 			next: None,
+ 		})
+ 	}
+
+	fn write(account: &T::AccountId, key: Option<T::KittyIndex>, item: KittyLinkedItem<T>) {
+ 		<OwnedKitties<T>>::insert(&(account.clone(), key), item);
+ 	}
+
+	pub fn append(account: &T::AccountId, kitty_id: T::KittyIndex) {
+		let head = Self::read_head(account);
+		let new_head = KittyLinkedItem {
+ 			prev: Some(kitty_id),
+ 			next: head.next,
+ 		};
+
+		Self::write_head(account, new_head);
+
+		let prev = Self::read(account, head.prev);
+		let new_prev = KittyLinkedItem {
+ 			prev: prev.prev,
+ 			next: Some(kitty_id),
+ 		};
+		Self::write(account, head.prev, new_prev);
+
+		let item = KittyLinkedItem {
+ 			prev: head.prev,
+ 			next: None,
+ 		};
+ 		Self::write(account, Some(kitty_id), item);
+	}
+
+	pub fn remove(account: &T::AccountId, kitty_id: T::KittyIndex) {
+		if let Some(item) = <OwnedKitties<T>>::take(&(account.clone(), Some(kitty_id))) {
+			let prev = Self::read(account, item.prev);
+			let new_prev = KittyLinkedItem {
+ 				prev: prev.prev,
+ 				next: item.next,
+ 			};
+
+			Self::write(account, item.prev, new_prev);
+
+			let next = Self::read(account, item.next);
+ 			let new_next = KittyLinkedItem {
+ 				prev: item.prev,
+ 				next: next.next,
+ 			};
+
+  			Self::write(account, item.next, new_next);
+		}
+	}
+}
+
+fn combine_dna(dna1: u8, dna2: u8, selector: u8) -> u8 {
+	((selector & dna1) | (!selector & dna2))
+}
+
+impl<T: Trait> Module<T> {
+	fn do_buy_kitty(sender: &T::AccountId, kitty_id: T::KittyIndex, max_price: BalanceOf<T>) -> Result {
+
+		ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+		let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+        ensure!(owner != *sender, "You can't buy your own cat");
+
+		let mut kitty = Self::kitty(kitty_id).unwrap();
+
+		let kitty_price = kitty.price;
+		ensure!(!kitty_price.is_zero(), "The cat you want to buy is not for sale");
```

Comment:
Is that better if we make the price to be option?

Reply:
optimize in  commit 4e97127e7a08e31db4c19796cdf78c591f6215fd